pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,459 @@
+	ctx context.Context
+	st  *cluster.Settings
+
+	tracer *tracing.Tracer
+}
+
+// NewRaftTracer creates a new RaftTracer with the given ambient context for the
+// replica.
+func NewRaftTracer(
+	ctx context.Context,
+	tracer *tracing.Tracer,
+	st *cluster.Settings,
+	numRegisteredStore *atomic.Int64,
+) *RaftTracer {
+	return &RaftTracer{ctx: ctx, tracer: tracer, st: st, numRegisteredStore: numRegisteredStore}
+}
+
+// MaxConcurrentRaftTraces is the maximum number of entries that can be traced
+// at any time on this store. Additional traces will be ignored until the number
+// of traces drops below the limit. Having too many active traces can negatively
+// impact performance as we iterate over all of them for some messages.
+//
+// TODO(baptist): Bump the default to a reasonable value like 10 that balances
+// usefulness with performance impact once we have validated the performance
+// impact.
+var MaxConcurrentRaftTraces = settings.RegisterIntSetting(
+	settings.SystemOnly,
+	"kv.raft.max_concurrent_traces",
+	"the maximum number of tracked raft traces, 0 will disable tracing",
+	0,
+	settings.IntInRange(0, 1000),
+)
+
+// reserveSpace checks if should register a new trace. If there are too many
+// registered traces it will not register and return false. The soft invariant
+// is that numRegisteredStore <= numAllowed which can be temporarily violated if
+// MaxConcurrentRaftTraces is lowered. This method will return true if we can
+// add one to the number registered for both the store and replica, otherwise it
+// will return false. This method is optimized for the `numAllowed == 0` case
+// and avoids loading `numRegisteredStore` until after this check.`
+func (r *RaftTracer) reserveSpace() bool {
+	numAllowed := MaxConcurrentRaftTraces.Get(&r.st.SV)
+	numRegisteredReplica := r.numRegisteredReplica.Load()
+
+	// This can only occur if the numAllowed setting has changed since a
+	// previous call to reserveSpace. If this happens flush all our current
+	// traces and don't register this request. Note that when this happens we
+	// also wont't log this request.
+	if numRegisteredReplica > numAllowed {
+		log.Infof(r.ctx, "flushing all traces due to setting change")
+		r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+			r.removeEntry(index)
+			return true
+		})
+		return false
+	}
+
+	if numAllowed == 0 {
+		return false
+	}
+
+	// The maximum number of traces has been reached for the store. We don't
+	// register tracing and return false.
+	numRegisteredStore := r.numRegisteredStore.Load()
+	if numRegisteredStore >= numAllowed {
+		return false
+	}
+
+	// Only increment the number of registered traces if the numRegistered
+	// hasn't changed. In the case of an ABA update, it does not break the
+	// invariant since some other trace was registered and deregistered, but
+	// there is still a slot available.
+	registerSucceeded := r.numRegisteredStore.CompareAndSwap(numRegisteredStore, numRegisteredStore+1)
```

Comment:
Could you also note that we can drop the trace here if someone outpaced us and incremented numRegisteredStore? And say that it's fine because it's a rare case.

Reply:
Added